source/auth/auth.rst

```diff
@@ -250,6 +250,48 @@ or "SCRAM-SHA-256").
 The cache entry value MUST be either the ``saltedPassword`` parameter or the
 combination of the ``clientKey`` and ``serverKey`` parameters.
 
+Speculative Authentication via Handshake
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:since: 4.4
+
+The ``isMaster`` handshake supports a new argument, ``speculativeSaslStart``,
+provided as a BSON object. Clients specifying this argument to ``isMaster`` will
+speculatively include the first ``saslStart`` command of an authentication handshake.
+This command may be provided to the server in parallel with any standard request for
+supported authentication mechanisms. This would permit clients to merge their
+``saslStart`` message with their ``isMaster`` request, and receive the ``saslStart``
+reply with the ``isMaster`` reply.
+
+If the speculatively issued command succeeds, the client should proceed with the next
+step of the exchange. If the command fails, the client should determine whether the
+``isMaster`` response contains enough information to select another mechanism.
```

Comment:
The next paragraph says older servers ignore `speculativeSaslStart`, so I assume the `isMaster` response won't include any bundled reply for auth. I wouldn't consider that a command failure, though, since `isMaster` itself will succeed.

Would "If the command fails" only apply on a 4.4+ server that processes `speculativeSaslStart` and reports a failure due to a typical authentication failure (as could happen today with the non-speculative conversation)? I think we should also clarify what happens if the speculative SASL conversation fails on 4.4+ for some reason and how that affects the wrapping `isMaster` response. Would the top-level `ok` field be `0`?

In the context of this paragraph, I think your second sentence is instructing drivers how to proceed on a pre-4.4 server. In that case, I'd suggest replacing "If the command fails" with "If the `isMaster` response does not include a `saslStart` reply..."

Reply:
If speculative authentication fails, that will not affect `isMaster`. The response will contain `ok: 1` but not contain a `speculativeAuthenticate` reply. I have added wording here to describe this.
I have replaced "if the command fails" with your suggested wording.